pkg/controller/manage/kmesh_manage.go

```diff
@@ -0,0 +1,275 @@
+				log.Infof("%s/%s: enable Kmesh manage", newPod.GetNamespace(), newPod.GetName())
+
+				nspath, _ := ns.GetPodNSpath(newPod)
+
+				if err := handleKmeshManage(nspath, true); err != nil {
+					log.Errorf("failed to enable Kmesh manage")
+					return
+				}
+				queue.AddRateLimited(QueueItem{pod: newPod, action: ActionAddAnnotation})
+			}
+
+			//delete Kmesh manage label for disable Kmesh control
+			if shouldEnroll(oldPod) && !shouldEnroll(newPod) {
+				log.Infof("%s/%s: disable Kmesh manage", newPod.GetNamespace(), newPod.GetName())
+
+				nspath, _ := ns.GetPodNSpath(newPod)
+				if err := handleKmeshManage(nspath, false); err != nil {
+					log.Errorf("failed to disable Kmesh manage")
+					return
+				}
+
+				queue.AddRateLimited(QueueItem{pod: newPod, action: ActionDeleteAnnotation})
+			}
+		},
+	}); err != nil {
+		return nil, fmt.Errorf("failed to add event handler to podInformer: %v", err)
+	}
+
+	return &KmeshManageController{
+		stopChan:        stopChan,
+		informerFactory: informerFactory,
+		podInformer:     podInformer,
+		queue:           queue,
+		client:          client,
+	}, nil
+}
+
+type KmeshManageController struct {
+	stopChan        chan struct{}
+	informerFactory informers.SharedInformerFactory
+	podInformer     cache.SharedIndexInformer
+	queue           workqueue.RateLimitingInterface
+	client          kubernetes.Interface
+}
+
+func (c *KmeshManageController) Run() {
+	go func() {
+		c.informerFactory.Start(c.stopChan)
+		if !cache.WaitForCacheSync(c.stopChan, c.podInformer.HasSynced) {
+			log.Error("Timed out waiting for caches to sync")
+			return
+		}
+		c.processItems()
+	}()
+}
+
+func (c *KmeshManageController) processItems() {
+	key, quit := c.queue.Get()
+	if quit {
+		return
+	}
+	defer c.queue.Done(key)
+
+	queueItem, ok := key.(QueueItem)
+	if !ok {
+		log.Errorf("expected QueueItem but got %T", key)
+		return
+	}
+
+	var err error
+	if queueItem.action == ActionAddAnnotation {
+		err = c.handleAddAnnotation(queueItem.pod)
+	} else if queueItem.action == ActionDeleteAnnotation {
+		err = c.handleDelAnnotation(queueItem.pod)
+	}
+
+	if err != nil {
+		if c.queue.NumRequeues(key) < MaxRetries {
+			log.Errorf("failed to handle pod %s/%s action %s, err: %v, will retry", queueItem.pod.Namespace, queueItem.pod.Name, queueItem.action, err)
+			c.queue.AddRateLimited(key)
+		} else {
+			log.Errorf("failed to handle pod %s/%s action %s after %d retries, err: %v, giving up", queueItem.pod.Namespace, queueItem.pod.Name, queueItem.action, MaxRetries, err)
+			c.queue.Forget(key)
+		}
+		return
+	}
+
+	c.queue.Forget(key)
+}
+
+func (c *KmeshManageController) handleAddAnnotation(pod *corev1.Pod) error {
+	return retry.OnError(retry.DefaultRetry, func(err error) bool {
+		return true
+	}, func() error {
+		return addKmeshAnnotation(c.client, pod)
+	})
+}
+
+func (c *KmeshManageController) handleDelAnnotation(pod *corev1.Pod) error {
+	return retry.OnError(retry.DefaultRetry, func(err error) bool {
```

Comment:
donot need to retry here as you have done it in the above caller

Reply:
> donot need to retry here as you have done it in the above caller

yes, this logic is redundant